### ISSUE TITLE
Drop implementation hangs

### ISSUE THREAD

**Issue by nrempel** [masalachai/epp-client#9]
I've been troubleshooting this issue but I haven't found a solution yet. It seems that the `EppClient` `Drop` implementation hangs on `logout`.

The process is waiting on the socket read: https://github.com/masalachai/epp-client/blob/c5fc82491b1b2ae3221045ce391a22c2ecf95ad9/epp-client/src/connection/registry.rs#L78

**Comment by nrempel** [comment]
I think they connection may have been closed by the server in this case which is why I'm reaching EOF. But it's strange that I am able to read the final <logout> response message if I call `logout()` explicitly.

**Comment by nrempel** [comment]
After troubleshooting further it seems that the EPP server is not writing any data to the socket. The reader is simply waiting for data. So far, I haven't discovered why this is the case. It's strange because I do get a response if I call `logout()` explicitly.

**Comment by masalachai** [comment]
If the server closed the connection, the client should run into an EOF error I think. I only have the connection to hexonet and I can't seem to reproduce this at the moment.

Can you let me know the calls you are making, If it's a bug with the data being read from the socket, the call prior to the logout could also have an issue, though you mentioned that the logout request seems to be written to the socket.

**Comment by nrempel** [comment]
In this case, to reproduce this, I'm simply creating a new EppClient and letting it go out of scope. The process never exits.

If I add an explicit call to logout, it succeeds and then logout is called a second time when EppClient goes out of scope which fails with "early eof" which indicates the socket was closed after the first successful logout.

**Comment by nrempel** [comment]
This won't help me understand _why_ this problem is occurring but a solution may be to remove the drop implementation for EppClient.

According to [rfc5734](https://datatracker.ietf.org/doc/html/rfc5734#section-2):

> An EPP session is normally ended by the client issuing an EPP
   <logout> command.  A server receiving an EPP <logout> command MUST
   end the EPP session and close the TCP connection with a CLOSE call.
   A client MAY end an EPP session by issuing a CLOSE call.

So it would be compliant to simply close the socket to end the session. I think this is actually a better design:

- Less work to be done in `Drop`
- Calling logout explicitly won't cause the program to panic

Basically, calling logout would be recommend but the caller would control this. In the case of programming error, the client would still be compliant as long as the socket closes cleanly.

What do you think about this approach?

**Comment by masalachai** [comment]
> In this case, to reproduce this, I'm simply creating a new EppClient and letting it go out of scope. The process never exits.

I tried this, and it didn't hang for me. Here are a couple of screenshots of the code and the output. It logged in and then logged out from the call from `Drop`.

> Calling logout explicitly won't cause the program to panic

Making the logout call explicit is alright, but I am unsure if it will ensure that this never happens again. I think it's hitting some kind of bug, because calling it explicitly shouldn't be all that different from calling it from `Drop` since the data being written to and read from the socket isn't differing in the two cases, so I think it mightt run into this problem where it hangs with some other registry or under other conditions even if we make the call explicit.

![epp-client-scope](https://user-images.githubusercontent.com/15325291/141154044-87fbf681-f292-419d-b4b6-147daaec1a05.png)
![epp-client-drop](https://user-images.githubusercontent.com/15325291/141154091-fdda5cdd-10fd-4ed0-b01a-6057dbdfc3ea.png)

**Comment by nrempel** [comment]
So I've tested this again with an example module like you have above and it actually works fine. I have been running this so far in an integration test using `#[tokio::test]` which still hangs.

Is there some reason why the behaviour would be different here? Maybe there is something in `#[tokio::test]` that I'm missing?

@djc do you have an idea why running this code in a test would hang?

**Comment by nrempel** [comment]
I've updated the test to use the following and it works:

```
#[tokio::test(flavor = "multi_thread", worker_threads = 10)]
```

[The default](https://docs.rs/tokio/1.13.0/tokio/attr.test.html#using-default) is for the `tokio::test` is `worker_threads = 1`. [The default](https://docs.rs/tokio/1.13.0/tokio/attr.main.html#multi-threaded-runtime) for `tokio::main` is multi-threaded.

**Comment by nrempel** [comment]
So I think using `block_on` in a single threaded executor is just a [Bad Idea™](https://github.com/tokio-rs/tokio/issues/2376#issuecomment-610371759). I haven't identified what exactly is happening but I think it is related the call to `block_on`.

In fact, if I change the `Drop` impl to [something equivalent](https://docs.rs/tokio/1.13.0/tokio/task/fn.block_in_place.html#examples) (I think) from tokio itself:

```rust

impl Drop for EppClient {
    fn drop(&mut self) {
        tokio::task::block_in_place(move || {
            tokio::runtime::Handle::current().block_on(async move {
                self.logout().await.unwrap();
            });
        });
    }
}
```

Then, the function panics with more information:

```
thread 'main' panicked at 'can call blocking only when running on the multi-threaded runtime', /home/nbrempel/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.13.0/src/runtime/thread_pool/worker.rs:286:21
```

So I think this might be expected behaviour when calling `block_on` from a single-threaded runtime.

**Comment by masalachai** [comment]
> So I think this might be expected behaviour when calling `block_on` from a single-threaded runtime.
> 
> In this case, I think we can safely remove the 2 `Drop` implementations. For the socket shutdown being performed for `EppConnection::drop`, I think we could move the shutdown call to the logout function to be called explicitly.

Yeah, doesn't seem like there is any way to use `Drop` with async, so having an explicit `close()` or something on the `EppClient` that also explicitly closes the `EppConnection` is the only option it seems.

Thanks for checking this so thoroughly.
